master/custom/factories.py

```diff
@@ -241,8 +241,14 @@ class UnixAsanBuild(UnixBuild):
     factory_tags = ["asan", "sanitizer"]
     # See https://bugs.python.org/issue42985 for more context on why
     # SIGSEGV is ignored on purpose.
-    compile_environ = {'ASAN_OPTIONS': 'detect_leaks=0:allocator_may_return_null=1:handle_segv=0'}
-    test_environ = {'ASAN_OPTIONS': 'detect_leaks=0:allocator_may_return_null=1:handle_segv=0'}
+    asan_options = 'detect_leaks=0:allocator_may_return_null=1:handle_segv=0'
+    compile_environ = {'ASAN_OPTIONS': asan_options}
+    test_environ = {
+        'ASAN_OPTIONS': asan_options,
+        # Workaround for a MacOS issue when ASAN is enabled
+        # python.exe(74602,0x7ff84626a700) malloc: nano zone abandoned due to inability to reserve vm space.
```

Comment:
there's no Python issue. I can file one, but I don't think it's actionable, and I haven't seen this issue popping up in any of my other Macs - only the AWS-hosted Intel Mac

Reply:
or did you mean linking to the issue in this repo? (https://github.com/python/buildmaster-config/issues/450)

Reply:
Please mention that test_cmd_line, test_posix and test_subprocess tests can fail because of this message.

This message doesn't look like a bug or a memory allocation failure. It's just a message related to an optimization which doesn't work in some cases.

Yeah, adding a link to `https://stackoverflow.com/questions/64126942/malloc-nano-zone-abandoned-due-to-inability-to-preallocate-reserved-vm-space` sounds like a good idea. And also to `https://github.com/python/buildmaster-config/issues/450`.

The more details the better for future readers.

Reply:
C code can be found at: https://opensource.apple.com/source/libmalloc/libmalloc-140.40.1/src/nano_malloc.c.auto.html

```c
#if NANO_PREALLOCATE_BAND_VM
	// Unconditionally preallocate the VA space set aside for nano malloc to
	// reserve it in all configurations. rdar://problem/33392283
	boolean_t preallocated = nano_preallocate_band_vm();
	if (!preallocated && _malloc_engaged_nano) {
		_malloc_printf(ASL_LEVEL_NOTICE, "nano zone abandoned due to inability to preallocate reserved vm space.\n");
		_malloc_engaged_nano = 0;
	}
#endif
```

Reply:
thanks @vstinner !

I'm thinking to abandon this change though, since I was able to set the env var on my worker (outside of buildbot, as @zware suggested), and the ASAN builder is now green.
Do you think it's still worthwhile to document this somewhere in the master configuration? (and if so, where?)
perhaps having the closed issue & PR as future reference would be good enough? or a note in the devguide?

Reply:
I'm not sure what's the best place to document this issue. Good to know that the buildbot is back to green 👍

Reply:
I replaced the workaround with comment block next to the builder.
let me know if you have feedback about the content or location of that comment!